java.util.NoSuchElementException: null

ObjectPool has 3 methods that currently not thread safe:
- futures in queue
- connections in use
- connections free

Currently, it is used for debugging and metrics.

They are not thread safe and might throw exceptions like the below:
```
Caused by: java.util.NoSuchElementException: null
	at java.util.WeakHashMap$HashIterator.nextEntry(WeakHashMap.java:808) [?:1.8.0_144]
	at java.util.WeakHashMap$KeyIterator.next(WeakHashMap.java:839) [?:1.8.0_144]
	at kotlin.collections.CollectionsKt___CollectionsKt.toList(_Collections.kt:1152) [kotlin-stdlib-1.3.10.jar:1.3.10-release-253 (1.3.10)]
	at com.github.jasync.sql.db.pool.ObjectPoolActor.getUsedItemsList(ActorBasedObjectPool.kt:211) [jasync-common-0.8.57-netty4.0.jar:?]
	at com.github.jasync.sql.db.pool.ActorBasedObjectPool.getUsedItems(ActorBasedObjectPool.kt:151) [jasync-common-0.8.57-netty4.0.jar:?]
	at com.github.jasync.sql.db.pool.ConnectionPool.inUse(ConnectionPool.kt:107) [jasync-common-0.8.57-netty4.0.jar:?]
	at com.outbrain.ob1k.db.experimental.MySqlConnectionPool.lambda$new$2(MySqlConnectionPool.java:49) [ob1k-db-1.1.146.jar:?]
```

--- a/db-async-common/src/main/java/com/github/jasync/sql/db/pool/ActorBasedObjectPool.kt
+++ b/db-async-common/src/main/java/com/github/jasync/sql/db/pool/ActorBasedObjectPool.kt
@@ -151,6 +151,9 @@ internal constructor(
     val usedItems: List<T> get() = actorInstance.usedItemsList
     val waitingForItem: List<CompletableFuture<T>> get() = actorInstance.waitingForItemList
     val usedItemsSize: Int get() = actorInstance.usedItemsSize
+    val waitingForItemSize: Int get() = actorInstance.waitingForItemSize
+    val availableItemsSize: Int get() = actorInstance.availableItemsSize
+
 }
 
 @Suppress("unused")
@@ -211,6 +214,8 @@ private class ObjectPoolActor<T : PooledObject>(
     val usedItemsList: List<T> get() = inUseItems.keys.toList()
     val waitingForItemList: List<CompletableFuture<T>> get() = waitingQueue.toList()
     val usedItemsSize: Int get() = inUseItems.size
+    val waitingForItemSize: Int get() = waitingQueue.size
+    val availableItemsSize: Int get() = availableItems.size
 
     var closed = false
 

--- a/db-async-common/src/main/java/com/github/jasync/sql/db/pool/ConnectionPool.kt
+++ b/db-async-common/src/main/java/com/github/jasync/sql/db/pool/ConnectionPool.kt
@@ -30,7 +30,7 @@ class ConnectionPool<T : Connection> @JvmOverloads constructor(
     private val executionContext: Executor = ExecutorServiceUtils.CommonPool
 ) : AsyncObjectPool<T>, Connection {
 
-    private val objectPool = ActorBasedObjectPool<T>(factory, configuration)
+    private val objectPool = ActorBasedObjectPool(factory, configuration)
 
     override val id: String
         get() = XXX("not implemented as it is not a real connection")
@@ -44,7 +44,7 @@ class ConnectionPool<T : Connection> @JvmOverloads constructor(
 
     override fun disconnect(): CompletableFuture<Connection> =
         if (this.isConnected()) {
-            objectPool.close().mapAsync(executionContext) { item -> this }
+            objectPool.close().mapAsync(executionContext) { this }
         } else {
             CompletableFuture.completedFuture(this)
         }
@@ -105,6 +105,21 @@ class ConnectionPool<T : Connection> @JvmOverloads constructor(
     fun queued(): List<CompletableFuture<T>> = objectPool.waitingForItem
 
     fun inUse(): List<T> = objectPool.usedItems
+
+    /**
+     * The number of connections that are currently in use for queries
+     */
+    val inUseConnectionsCount: Int get() = objectPool.usedItemsSize
+
+    /**
+     * The number of futures that were submitted by not yet assigned to a connection from the pool
+     */
+    val futuresWaitingForConnectionCount: Int get() = objectPool.waitingForItemSize
+
+    /**
+     * The number of connections that are idle and ready to use
+     */
+    val idleConnectionsCount: Int get() = objectPool.availableItemsSize
 
     override fun take(): CompletableFuture<T> = objectPool.take()
 

--- a/db-async-common/src/test/java/com/github/jasync/sql/db/pool/ActorBasedObjectPoolTest.kt
+++ b/db-async-common/src/test/java/com/github/jasync/sql/db/pool/ActorBasedObjectPoolTest.kt
@@ -225,6 +225,8 @@ class ActorBasedObjectPoolTest {
         Thread.sleep(1000)
         System.gc()
         await.untilCallTo { tested.usedItemsSize } matches { it == 0 }
+        await.untilCallTo { tested.waitingForItemSize } matches { it == 0 }
+        await.untilCallTo { tested.availableItemsSize } matches { it == 0 }
         System.gc() //to show leak in logging
         Thread.sleep(1000)
     }

--- a/postgresql-async/src/test/java/com/github/aysnc/sql/db/integration/pool/SingleThreadedAsyncObjectPoolSpec.kt
+++ b/postgresql-async/src/test/java/com/github/aysnc/sql/db/integration/pool/SingleThreadedAsyncObjectPoolSpec.kt
@@ -44,6 +44,9 @@ class SingleThreadedAsyncObjectPoolSpec : DatabaseTestHelper() {
             assertThat(pool.availables().size).isEqualTo(0)
             assertThat(pool.inUse().size).isEqualTo(1)
             assertThat(pool.queued().size).isLessThanOrEqualTo(3)
+            assertThat(pool.idleConnectionsCount).isEqualTo(0)
+            assertThat(pool.inUseConnectionsCount).isEqualTo(1)
+            assertThat(pool.futuresWaitingForConnectionCount).isLessThanOrEqualTo(3)
 
             /* pool.take call checkout that call this.mainPool.action,
             so enqueuePromise called in executorService,

